src/main/kotlin/ecommerce/entities/Product.kt

```diff
@@ -16,11 +22,59 @@ class Product(
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     val id: Long? = null,
     @Column(name = "name", nullable = false)
-    val name: String,
+    var name: String,
     @Column(name = "price", nullable = false)
-    val price: Double,
+    var price: Double,
     @Column(name = "image_url", nullable = false)
-    val imageUrl: String,
+    var imageUrl: String,
     @OneToMany(mappedBy = "product", cascade = [CascadeType.ALL], orphanRemoval = true)
     val cartItems: Set<CartItem> = emptySet(),
-)
+    @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true)
+    @JoinColumn(name = "product", nullable = false)
+    private val _options: MutableList<Option> = mutableListOf(),
+) {
+    var options: List<Option>
+        get() = _options.toList()
+        set(value) {
+            _options.clear()
+            _options.addAll(value)
+        }
+
+    fun addOption(option: Option) {
+        if (_options.any { it.name == option.name }) {
+            throw InvalidOptionNameException("Option with name '${option.name}' already exists")
+        }
+        _options.add(option)
+    }
+
+    fun copyFrom(
+        other: ProductRequestDTO,
+        optionDTOs: Set<OptionDTO> = emptySet(),
+    ) {
+        this.name = other.name
+        this.price = other.price
+        this.imageUrl = other.imageUrl
+        this.options = mapOptionDTOs(optionDTOs)
+    }
+
+    fun copyFrom(
+        other: ProductPatchDTO,
+        optionDTOs: Set<OptionDTO> = emptySet(),
+    ) {
+        other.name?.let { if (it.isNotBlank()) this.name = it }
+        other.price?.let { this.price = it }
+        other.imageUrl?.let { if (it.isNotBlank()) this.imageUrl = it }
+        if (optionDTOs.isNotEmpty()) {
+            this.options = mapOptionDTOs(optionDTOs)
+        }
+    }
+
+    private fun mapOptionDTOs(optionDTOs: Set<OptionDTO>): List<Option> {
+        return optionDTOs.map { dto ->
+            _options.find { it.id == dto.id }?.apply {
+                name = dto.name
+                quantity = dto.quantity
+            } ?: dto.toEntity(this)
+        }
+    }
```

Comment:
It doesn’t seem ideal for the domain model to depend on a DTO. When the flow is one-directional in the layer structure, it becomes easier to manage dependencies. It would be better if this conversion happens externally. 🙂